src/java.base/share/classes/java/util/Collections.java

```diff
@@ -1164,6 +1170,81 @@ public static <T> Set<T> unmodifiableSet(Set<? extends T> s) {
         UnmodifiableSet(Set<? extends E> s)     {super(s);}
         public boolean equals(Object o) {return o == this || c.equals(o);}
         public int hashCode()           {return c.hashCode();}
+
+        @java.io.Serial
+        private Object readResolve() {
+            if (c instanceof RegularEnumSetCompatible<?> es) {
+                return new UnmodifiableRegularEnumSet<>(es);
+            }
+            if (c instanceof JumboEnumSetCompatible<?> es) {
+                return new UnmodifiableJumboEnumSet<>(es);
+            }
+            return this;
+        }
+    }
+
+    /**
+     * @serial include
+     */
+    static final class UnmodifiableRegularEnumSet<E extends Enum<E>> extends UnmodifiableSet<E>
+            implements RegularEnumSetCompatible<E> {
+        @java.io.Serial
+        private static final long serialVersionUID = -1110577510253015312L;
+
+        final RegularEnumSetCompatible<? extends E> es;
```

Comment:
Can't you just use `<E>` as the type argument instead, for all occurrences? Can avoid a cast in `elementType()` too, and it's compatible argument to superclass constructor.